esp/services/ws_workunits/ws_wudetails.cpp

```diff
@@ -0,0 +1,349 @@
+
+    DBGLOG("Id list count: %d", ids.ordinality());
+    DBGLOG("ReturnAttribute  list count: %d", returnAttribList.ordinality());
+    DBGLOG("filterNestedScopeTypes count: %d", nestedScopeTypes.ordinality());
+    DBGLOG("AttributeFilter count: %d", arrayAttributeFilter.ordinality());
+
+    ids.sortAscii();      // Will need to lookup up 'id' in ids (for bSearch)
+    IArrayOf<IEspWUResponseAttribute> attribs;
+    SCMStringBuffer currentScope,currentScopeType,currentId;
+    unsigned statsCount = 0, statsAdded = 0;
+    ForEach(statsIter)
+    {
+        ++statsCount;
+        IConstWUStatistic & wuStat = statsIter.query();
+
+        SCMStringBuffer wuScope,  wuId;
+        wuStat.getScope(wuScope);
+        getScopeIdFromScopeName(wuId, wuScope);
+
+        if (limitToIdList && ids.bSearch(wuId.str(),funcCompareString)==NotFound)
+              continue;
+
+        if (maxDepth && getScopeDepth(wuScope) > maxDepth)
+            continue;
+
+        if (filterNestedScopeTypes && nestedScopeTypes.bSearch(wuStat.getScopeType(),funcCompareUnsignedShort)==NotFound)
+              continue;
+
+        if (filterByAttributeValue)
+        {
+            tmpForAttribSearch->setStatisticKind(wuStat.getKind());
+            aindex_t pos = arrayAttributeFilter.bSearch(tmpForAttribSearch,funcCompareAttribFilter);
+            if (pos!=NotFound)
+            {
+                CAttributeFilter & attributeFilter = arrayAttributeFilter.item(pos);
+                unsigned __int64 graphRawValue = wuStat.getValue();
+
+                if (attributeFilter.hasExactValue())
+                {
+                    if (graphRawValue != attributeFilter.getExactValue()) continue;
+                }
+                else
+                {
+                    if (attributeFilter.hasMinValue() && graphRawValue < attributeFilter.getMinValue()) continue;
+                    if (attributeFilter.hasMaxValue() && graphRawValue > attributeFilter.getMaxValue()) continue;
+                }
+            }
+        }
+        // Check if scope name has changed. If so, need to return the list attributes for previous scope
+        if (strcmp(currentScope.str(),wuScope.str())!=0)
+        {
+            if (currentScope.length()>0 && attribs.ordinality()>0)
+            {
+                Owned<IEspWUScope> respScope = buildRespWUScope(currentScope, currentScopeType, currentId, attribs);
+                respScopes.append(*respScope.getClear());
+                attribs.clear();
+            }
+            currentScope.set(wuScope.str());
+            currentScopeType.set(queryScopeTypeName(wuStat.getScopeType()));
+            currentId.set(wuId.str());
+        }
+
+        if (returnSpecifyReturnAttribs)
+        {
+            const unsigned sk = (unsigned) wuStat.getKind();
+            if (returnAttribList.bSearch(sk,funcCompareUnsignedShort)==NotFound)
+               continue;
+        }
+
+        if (wuStat.getTimestamp()<minVersion)
```

Comment:
I would cache wuStat.getTimeStamp in a lcoal variable, since you do not know how expensive it is to return it.